File: leetcode/rust-leetcode/src/14.longest-common-prefix.rs
impl Solution {
    pub fn longest_common_prefix(mut strs: Vec<String>) -> String {
        let mut lcp = String::new();
        strs.sort_by_key(|s| s.len());
        for i in 0..strs[0].len() {
            let c = strs[0].get(i..i+1).unwrap();
            let mut flag = true;
            for j in 1..strs.len() {
                if c != strs[j].get(i..i+1).unwrap() {
                    flag = false;
                    break;
                }
            }
            if flag {
                lcp.push_str(c);
            } else {
                break;
            }
        }
        lcp
    }
}

/*
*/

impl Solution {
    pub fn longest_common_prefix(strs: Vec<String>) -> String {
        if strs.is_empty() {
            return "".to_string();
        }

        let char_vecs: Vec<Vec<char>> = strs.iter().map(|s| s.chars().collect()).collect();
        let mut prefix = String::new();
        
        // Find the length of the shortest string
        let min_length = char_vecs.iter().map(|vec| vec.len()).min().unwrap_or(0);

        'outer: for i in 0..min_length {
            let c = char_vecs[0][i];  // First string's current character for comparison
            for s in &char_vecs[1..] {  // Start from second string
                if i >= s.len() || s[i] != c {
                    // Break the outer loop if character does not match or index is out of bounds
                    break 'outer;
                }
            }
            prefix.push(c);  // Add character to prefix if it matches all strings so far
        }

        prefix
    }
}

File: leetcode/cpp/14.longest-common-prefix.cpp
class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
        if (strs[0].length() == 0) {
            return "";
        }
        int i = 0;
        char ref;
        bool done = false;
        while (!done && i < strs[0].length()) {
            ref = strs[0][i];
            for (string s: strs) {
                if (s.length() == i || s[i] != ref) {
                    done = true;
                    break;
                }
            }
            if (!done) {i++;}
        }
        return strs[0].substr(0, i);
    }
};

/*  */

class Solution {
public:
    string longestCommonPrefix(vector<string>& v) {
        string ans="";
        sort(v.begin(),v.end());
        int n=v.size();
        string first=v[0],last=v[n-1];
        for(int i=0;i<min(first.size(),last.size());i++){
            if(first[i]!=last[i]){
                return ans;
            }
            ans+=first[i];
        }
        return ans;
    }
};

